refactor: remove Optional & nullopt

fanquake opened the pull request:
Same rationale & motivation as #21404, which turned out to be quite low in the number of potential conflicts. Lets see what the bot has to say here.

jnewbery commented on src/node/psbt.h:
```diff
@@ -25,18 +25,18 @@ struct PSBTInputAnalysis {
+    std::optional<size_t> estimated_vsize;      //!< Estimated weight of the transaction
+    std::optional<CFeeRate> estimated_feerate;  //!< Estimated feerate (fee / weight) of the transaction
+    std::optional<CAmount> fee;                 //!< Amount of fee being paid by the transaction
     std::vector<PSBTInputAnalysis> inputs; //!< More information about the individual inputs of the transaction
```
You could realign these comments if you wanted to be nice.

MarcoFalke replied:
run a scripted diff with clang format on the diff? (*hides)

jnewbery commented on src/net.cpp:
```diff
@@ -215,7 +214,7 @@ Optional<CAddress> GetLocalAddrForPeer(CNode *pnode)
     }
     // Address is unroutable. Don't advertise.
-    return nullopt;
+    return std::nullopt;
```
It may be slightly more idiomatic to replace all the `return std::nullopt;` with `return {};`

jnewbery commented on src/net.cpp:
```diff
@@ -747,12 +746,12 @@ Optional<CNetMessage> V1TransportDeserializer::GetMessage(const std::chrono::mic
                  m_node_id);
         out_err_raw_size = msg->m_raw_message_size;
-        msg = nullopt;
+        msg = std::nullopt;
```
Perhaps `msg.reset()` is more idiomatic.

jnewbery commented on src/txmempool.cpp:
```diff
@@ -890,11 +889,11 @@ const CTransaction* CTxMemPool::GetConflictTx(const COutPoint& prevout) const
     auto it = mapTx.find(txid);
     if (it != mapTx.end()) return it;
-    return Optional<txiter>{};
+    return std::optional<txiter>{};
```
```suggestion
    return {};
```

fanquake replied:
This should also finally quell the `-Wmaybe-uninitialized ` false positives we've seen. i.e #21318, #21248.

jnewbery commented on src/validation.h:
```diff
@@ -208,7 +207,7 @@ struct MempoolAcceptResult {
     explicit MempoolAcceptResult(TxValidationState state)
         : m_result_type(ResultType::INVALID),
-        m_state(state), m_replaced_transactions(nullopt), m_base_fees(nullopt) {
+        m_state(state), m_replaced_transactions(std::nullopt), m_base_fees(std::nullopt) {
```
No need for these initializers. `std::optional` default initializes to `std::nullopt`

jnewbery commented on src/wallet/rpcdump.cpp:
```diff
@@ -1788,7 +1788,7 @@ RPCHelpMan listdescriptors()
         spk.pushKV("active", active);
         const auto& type = wallet_descriptor.descriptor->GetOutputType();
-        if (active && type != nullopt) {
+        if (active && type != std::nullopt) {
```
More idiomatic:

```suggestion
        if (active && type) {
```

jnewbery commented on src/wallet/rpcwallet.cpp:
```diff
@@ -219,7 +218,7 @@ static void SetFeeEstimateMode(const CWallet& wallet, CCoinControl& cc, const Un
         if (override_min_fee) cc.fOverrideFeeRate = true;
         // Default RBF to true for explicit fee_rate, if unset.
-        if (cc.m_signal_bip125_rbf == nullopt) cc.m_signal_bip125_rbf = true;
+        if (cc.m_signal_bip125_rbf == std::nullopt) cc.m_signal_bip125_rbf = true;
```
```suggestion
        if (!cc.m_signal_bip125_rbf) cc.m_signal_bip125_rbf = true;
```

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -84,10 +83,10 @@ bool RemoveWalletSetting(interfaces::Chain& chain, const std::string& wallet_nam
                                 std::vector<bilingual_str>& warnings)
 {
-    if (load_on_startup == nullopt) return;
+    if (load_on_startup == std::nullopt) return;
```
```suggestion
    if (!load_on_startup) return;
```

jnewbery commented on src/miner.cpp:
```diff
@@ -96,8 +96,8 @@ void BlockAssembler::resetBlock()
 
-Optional<int64_t> BlockAssembler::m_last_block_num_txs{nullopt};
-Optional<int64_t> BlockAssembler::m_last_block_weight{nullopt};
+std::optional<int64_t> BlockAssembler::m_last_block_num_txs{std::nullopt};
```
Not required. `std::optional<>` will default initialize to `std::nullopt`. These lines can be removed entirely.

MarcoFalke replied:
They are static, but this should still zero initialize: https://en.cppreference.com/w/cpp/language/zero_initialization

jnewbery replied:
Does this seem right to you, @MarcoFalke:

```diff
diff --git a/src/miner.cpp b/src/miner.cpp
index 56543e743e..5239836e0f 100644
--- a/src/miner.cpp
+++ b/src/miner.cpp
@@ -96,8 +96,8 @@ void BlockAssembler::resetBlock()
     nFees = 0;
 }
 
-std::optional<int64_t> BlockAssembler::m_last_block_num_txs{std::nullopt};
-std::optional<int64_t> BlockAssembler::m_last_block_weight{std::nullopt};
+/* std::optional<int64_t> BlockAssembler::m_last_block_num_txs{std::nullopt}; */
+/* std::optional<int64_t> BlockAssembler::m_last_block_weight{std::nullopt}; */
 
 std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock(const CScript& scriptPubKeyIn)
 {
diff --git a/src/miner.h b/src/miner.h
index 286a8c693e..9dadbf1507 100644
--- a/src/miner.h
+++ b/src/miner.h
@@ -159,8 +159,8 @@ public:
     /** Construct a new block template with coinbase to scriptPubKeyIn */
     std::unique_ptr<CBlockTemplate> CreateNewBlock(const CScript& scriptPubKeyIn);
 
-    static std::optional<int64_t> m_last_block_num_txs;
-    static std::optional<int64_t> m_last_block_weight;
+    inline static std::optional<int64_t> m_last_block_num_txs{};
+    inline static std::optional<int64_t> m_last_block_weight{};
 
 private:
     // utility functions
```

MarcoFalke replied:
Is the `inline` needed? Any version that passes mining_basic.py under valgrind should be correct. See fa178a6385bf300499fb18940051fc4142fb5b6b

jnewbery replied:
I'm pretty sure inline is required:

```
In file included from miner.cpp:6:
./miner.h:162:35: error: non-const static data member must be initialized out of line
    static std::optional<int64_t> m_last_block_num_txs{};
                                  ^~~~~~~~~~~~~~~~~~~~~~
./miner.h:163:35: error: non-const static data member must be initialized out of line
    static std::optional<int64_t> m_last_block_weight{};
                                  ^~~~~~~~~~~~~~~~~~~~~
2 errors generated.
```

Inline member variables have been possible since c++17: https://www.codingame.com/playgrounds/2205/7-features-of-c17-that-will-simplify-your-code/inline-variables.

MarcoFalke replied:
thanks. Seems fine to change this to `inline static` initialization.

jnewbery reviewed (commented):
ACK 8d2046678556294a97e2e9809e55c9ce19860569.

Change looks correct, but now that we're using `std::optional` rather than `boost::optional`, we can be a lot more idiomatic. `std::nullopt` almost never needs to be used explicitly - `std::optional<>` objects are default initialized to `std::nullopt`, `std::nullopt` can be initialized with empty braces `{ }`, and the boolean operator can be used instead of explicit comparison with `std::nullopt`.

JeremyRubin commented:
might be worth changing the style guide -- I personally think that using nullopt is a good tradeoff between verbosity and legibility ie https://stackoverflow.com/questions/62933403/why-do-we-need-stdnullopt in many cases.

jonatack commented:
> might be worth changing the style guide -- I personally think that using nullopt is a good tradeoff between verbosity and legibility ie https://stackoverflow.com/questions/62933403/why-do-we-need-stdnullopt in many cases.

Similar questions in a couple places today: https://github.com/bitcoin/bitcoin/pull/21407#discussion_r592261547

practicalswift commented:
Concept ACK

Thanks for removing old cruft!

glozow reviewed (commented):
why no `s/#include <optional.h>/#include <optional>` ?

fanquake committed:
scripted-diff: remove Optional & nullopt

-BEGIN VERIFY SCRIPT-
git rm src/optional.h

sed -i -e 's/Optional</std::optional</g' $(git grep -l 'Optional<' src)

sed -i -e 's/{nullopt}/{std::nullopt}/g' $(git grep -l 'nullopt' src)
sed -i -e 's/ nullopt;/ std::nullopt;/g' $(git grep -l 'nullopt' src)
sed -i -e 's/ nullopt)/ std::nullopt)/g' $(git grep -l 'nullopt' src)
sed -i -e 's/(nullopt)/(std::nullopt)/g' $(git grep -l 'nullopt' src)
sed -i -e 's/ nullopt,/ std::nullopt,/g' $(git grep -l 'nullopt' src)
sed -i -e 's/? nullopt :/? std::nullopt :/g' $(git grep -l 'nullopt' src)
sed -i -e 's/: nullopt}/: std::nullopt}/g' $(git grep -l 'nullopt' src)

sed -i -e '/optional.h \\/d' src/Makefile.am

sed -i -e '/#include <optional.h>/d' src/test/fuzz/autofile.cpp src/test/fuzz/buffered_file.cpp src/test/fuzz/node_eviction.cpp

sed -i -e 's/#include <optional.h>/#include <optional>/g' $(git grep -l '#include <optional.h>' src)
-END VERIFY SCRIPT-

fanquake committed:
refactor: post Optional<> removal cleanups

fanquake commented:
[Similar to #21404](https://github.com/bitcoin/bitcoin/pull/21404#issuecomment-797163202) I've left comments where new `Optional<>` usage is being introduced. i.e [here](https://github.com/bitcoin/bitcoin/pull/16037/files#r594007135), [here](https://github.com/bitcoin/bitcoin/pull/16545/files#r594006161), [here](https://github.com/bitcoin/bitcoin/pull/18554/files#r594005916) etc.

> why no s/#include <optional.h>/#include <optional> ?

It was originally this way, but the duplicate include linter was unhappy. This should now be fixed.

Have added some additional commits to take post-removal suggestions.

jnewbery commented on src/interfaces/chain.h:
```diff
@@ -5,7 +5,7 @@
 #define BITCOIN_INTERFACES_CHAIN_H
 
-#include <optional.h>               // For Optional and nullopt
+#include <optional>               // For Optional and nullopt
```
Remove comment (and separate from project includes)

MarcoFalke replied:
style nit: :eye: Comment wrong. Also, would be nice if you sorted the includes

jnewbery commented on src/validation.h:
```diff
@@ -206,7 +206,7 @@ struct MempoolAcceptResult {
     explicit MempoolAcceptResult(TxValidationState state)
         : m_result_type(ResultType::INVALID),
-        m_state(state), m_replaced_transactions(nullopt), m_base_fees(nullopt) {
+        m_state(state), m_replaced_transactions(), m_base_fees() {
```
Just remove these from the initializer list entirely:

```
    explicit MempoolAcceptResult(TxValidationState state)
        : m_result_type(ResultType::INVALID), m_state(state)
        { Assume(!state.IsValid()); } // Can be invalid or error
```

jnewbery reviewed (commented):
Just a couple more style comments

fanquake commented:
I've added new includes, sorted existing includes, removed a comment and realigned other comments.

jnewbery commented:
ACK dca584e768

practicalswift commented on src/net.cpp:
```diff
@@ -747,12 +747,12 @@ Optional<CNetMessage> V1TransportDeserializer::GetMessage(const std::chrono::mic
                  m_node_id);
         out_err_raw_size = msg->m_raw_message_size;
-        msg = nullopt;
+        msg = std::nullopt;
```
`msg.reset();` is now used on L755. Was the intention to use it also on L750?

practicalswift commented:
cr ACK dca584e7684c38afb9e5267104d0c0a008fd7477: patch looks correct. The AppVeyor failure looks spurious.

ryanofsky commented on src/miner.h:
```diff
@@ -160,8 +160,8 @@ class BlockAssembler
 
-    static std::optional<int64_t> m_last_block_num_txs;
-    static std::optional<int64_t> m_last_block_weight;
+    inline static std::optional<int64_t> m_last_block_num_txs{};
+    inline static std::optional<int64_t> m_last_block_weight{};
```
At first glance, I thought adding inline keyword here might cause separate `m_last_block_num_txs` and `m_last_block_weight` variable instances to exist in different translation units (different variables with the same name `miner.cpp` and `rpc/mining.cpp` that would lead to bugs), but apparently there is new linker magic in c++17 to prevent this!

jnewbery replied:
✨ m a g i c ✨

laanwj replied:
Good to know this

jnewbery replied:
(more background here: https://github.com/bitcoin/bitcoin/pull/21415#discussion_r592546422)

ryanofsky reviewed (approved):
Code review ACK dca584e7684c38afb9e5267104d0c0a008fd7477. Kill it with fire 🔥🧨🧑‍🚒

laanwj commented:
> I personally think that using nullopt is a good tradeoff between verbosity and legibility ie https://stackoverflow.com/questions/62933403/why-do-we-need-stdnullopt in many cases.

I have the same opinion about this. Using `std::optional` is a no-brainer but replacing `nullopt` with `{}` doesn't always seem like a readability improvement to me. `{}` is this kind of wildcard that can be anything from an empty structure, a default initializer, etc, `nullopt` is type-safe and clear.

fanquake commented:
> I have the same opinion about this.

Ok. I've  dropped f603e792c083fe9c9072dc3e753dcf9922583ee0.

practicalswift commented:
cr ACK ebc4ab721b0371c0ef217c0f5bd7d42613e951e6: patch looks correct

jnewbery commented:
utACK ebc4ab721b0371c0ef217c0f5bd7d42613e951e6

laanwj commented:
Code review ACK ebc4ab721b0371c0ef217c0f5bd7d42613e951e6

jonatack reviewed (commented):
Posthumous ACK ebc4ab721b037
